### PR TITLE
Distinct in order optimization

### PR DESCRIPTION
**Pull request by dimarub2000**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Performance Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Distinct containing primary key columns is now optimized.

Detailed description / Documentation draft:
TBA

### CODE REVIEW

File: src/Storages/ReadInOrderOptimizer.cpp [hunk 1]
```diff
@@ -12,6 +12,27 @@ namespace ErrorCodes
     extern const int LOGICAL_ERROR;
 }
 
+static std::optional<Names> tryGetSortingKeyColumns(const StoragePtr & storage)
+{
+    Names sorting_key_columns;
+    if (const auto * merge_tree = dynamic_cast<const MergeTreeData *>(storage.get()))
```
Comment (CurtizJ) [inline src/Storages/ReadInOrderOptimizer.cpp:18]: Now `getSortingKeyColumns` method is a part of `IStorage` interface and it returns empty list of columns if storage doesn't support sorting key. So, dynamic_casts are unneeded.

File: src/Storages/ReadInOrderOptimizer.h [hunk 2]
```diff
@@ -29,4 +29,14 @@ class ReadInOrderOptimizer
...
 };
 
+class ReadInOrderOptimizerForDistinct
+{
+public:
+    explicit ReadInOrderOptimizerForDistinct(const Names & source_columns_);
+    InputOrderInfoPtr getInputOrder(const StoragePtr & storage) const;
+
+private:
+    Names source_columns;
```
Comment (CurtizJ) [inline src/Storages/ReadInOrderOptimizer.h:39]: It's better to use `NameSet`.

File: src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp [hunk 3]
```diff
@@ -637,9 +637,23 @@ Pipes MergeTreeDataSelectExecutor::readFromParts(
...
+        else
+            prefer_distinct_order = true;
+        if (query_info.distinct_order_info)
+            distinct_prefix_size = query_info.distinct_order_info->order_key_prefix_descr.size();
+        if (!settings.optimize_aggregation_in_order && distinct_prefix_size > prefix_size)
+        {
+            prefer_distinct_order = true;
+            prefix_size = distinct_prefix_size;
+        }
+        std::cerr << std::boolalpha << prefer_distinct_order << "\n";
```
Comment (CurtizJ) [inline src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp:656]: Maybe it's better to do it at previous level (`InterpreterSelectQuery` or `SyntaxAnalyzer`)?

File: src/Storages/ReadInOrderOptimizer.cpp [hunk 4]
```diff
@@ -125,4 +132,29 @@ InputOrderInfoPtr ReadInOrderOptimizer::getInputOrder(const StoragePtr & storage
...
+
+    Names sorting_key_columns = *sorting_key_object;
+    SortDescription order_key_prefix_descr;
+    for (const auto & sorting_key_column: sorting_key_columns)
+    {
+        const auto it = std::find(source_columns.cbegin(), source_columns.cend(), sorting_key_column);
+        if (it == source_columns.cend())
+            break;
+        order_key_prefix_descr.emplace_back(sorting_key_column, 1, 1);
+    }
```
Comment (CurtizJ) [inline src/Storages/ReadInOrderOptimizer.cpp:154]: Maybe also support monotonic functions here?

File: src/Processors/Transforms/DistinctSortedTransform.cpp [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+    /// Calculate not-sorted columns positions
+    other_columns_pos.reserve(source_columns.size());
+    for (const auto & source_column : source_columns)
+    {
+        size_t pos = header_.getPositionByName(source_column);
+        if (std::find(sorted_columns_pos.begin(), sorted_columns_pos.end(), pos) != sorted_columns_pos.end())
+            continue;
+
+        const auto & col = header_.getByPosition(pos).column;
+        if (!(col && isColumnConst(*col)))
```
Comment (CurtizJ) [inline src/Processors/Transforms/DistinctSortedTransform.cpp:59]: What if `col` will be `nullptr`.  It's impossible, but technically condition is not correct. Looks like it shoud be `col && !isColumnConst(*col).

**Comment by azat** [comment]
@dimarub2000 worthwhile patch, are you planning to finish this?

**Comment by alexey-milovidov** [comment]
@dimarub2000 This is a very good feature. We will find developers to finish this PR.

**Comment by devcrafter** [comment]
1st version is done in #37803. For further improvements, see #39313

**Pull request closed without merging**